pubsub/api/PubsubTest.cs

```diff
@@ -0,0 +1,445 @@
+                subscriptionId);
+            subscriber.DeleteSubscription(subscriptionName);
+            // [END delete_subscription]
+        }
+
+        public void DeleteTopic(string topicId, PublisherClient publisher)
+        {
+            // [START delete_topic]
+            string topicName = PublisherClient.FormatTopicName(_projectId, 
+                topicId);
+            publisher.DeleteTopic(topicName);
+            // [END delete_topic]
+        }
+
+        // [START retry]
+        /// <summary>
+        /// Retry the action when a Grpc.Core.RpcException is thrown.
+        /// </summary>
+        private T RetryRpc<T>(Func<T> action)
+        {
+            List<Grpc.Core.RpcException> exceptions = null;
+            var delayMs = _retryDelayMs;
+            for (int tryCount = 0; tryCount < _retryCount; ++tryCount)
+            {
+                try
+                {
+                    return action();
+                }
+                catch (Grpc.Core.RpcException e)
+                {
+                    if (exceptions == null)
+                        exceptions = new List<Grpc.Core.RpcException>();
+                    exceptions.Add(e);
+                }
+                System.Threading.Thread.Sleep(delayMs);
+                delayMs *= 2;  // Exponential back-off.
+            }
+            throw new AggregateException(exceptions);
+        }
+
+        private void RetryRpc(Action action)
+        {
+            try {
+                RetryRpc(() => { action(); return 0; });
+            }
+            catch (AggregateException exceptions)
+            {
+                throw exceptions;
+            }
+        }
+        // [END retry]
+
+        public void RpcRetry(string topicId, string subscriptionId,
+            PublisherClient publisher, SubscriberClient subscriber)
+        {
+            string topicName = PublisherClient.FormatTopicName(_projectId, topicId);
+            string subscriptionName =
+                SubscriberClient.FormatSubscriptionName(_projectId, subscriptionId);
+            try
+            {
+                RetryRpc(() =>
+                {
+                    // Subscribe to Topic
+                    // This should fail since Topic has not yet been created
+                    subscriber.CreateSubscription(subscriptionName, topicName,
+                        pushConfig: null, ackDeadlineSeconds: 60);
+                });
+            }
+            catch (AggregateException exceptions)
+            {
+                if (exceptions.InnerExceptions.Count() == _retryCount)
+                {
+                    // [START retry]
+                    try
+                    {
+                        RetryRpc(() =>
+                        {
+                            //Create Topic
+                            publisher.CreateTopic(topicName);
+                        });
+                    }
+                    catch (AggregateException ae)

```

Comment:
If the topic already exists, this code will try to create it 3 times.  Instead, it should observe the StatusCode=AlreadyExists in the first exception and not try twice more.

Also, I think the API could be better here.  We should not have to parse text strings to observe this.  Text strings notoriously change from version to version, and are generally fragile.  File a bug on https://github.com/GoogleCloudPlatform/google-cloud-dotnet and ask for a better API.